docs/learn/prompts.md

```diff
@@ -1,219 +1,425 @@
 
 ```python
+from mirascope.core import BasePrompt, prompt_template
 from pydantic import computed_field
 
+
+@prompt_template("Recommend a {genre} {book_type} for a {age_group} reader.")
 class BookRecommendationPrompt(BasePrompt):
-	"""Recommend a {genre} {book_type}."""
-		
-  genre: str
-  
-  @computed_field
-  def book_type(self) -> str:
-	  if self.genre in ["science", "history", "math"]:
-		  return "textbook"
-		return "book"
-		
-prompt = BookRecommendationPrompt(genre="fantasy")
-print(prompt)
-# > Recommend a fantasy book.
+    genre: str
+    age_group: str
 
-prompt = BookRecommendationPrompt(genre="science")
+    @computed_field
+    @property
+    def book_type(self) -> str:
+        if self.age_group == "child":
+            return "picture book"
+        elif self.age_group == "young adult":
+            return "novel"
+        else:
+            return "book"
+
+
+prompt = BookRecommendationPrompt(genre="fantasy", age_group="child")
 print(prompt)
-# > Recommend a science textbook.
+# > Recommend a fantasy picture book for a child reader.
 ```
 
-Using a `computed_field` will include the computed field in the dump, which is useful for seeing what affects your prompts.
+!!! tip "Retrieving External Data"
+
+    Computed fields can be particularly useful when you want to inject information you retrieve from an external source. For example, perhaps you want to retrieve context from a document store to augment the generation with relevant information.
+
+### Metadata
+
+You can add metadata to your prompts using the `@metadata` decorator. This will attach a `Metadata` object, which is a simple `TypedDict` with a single typed key `tags: set[str]`.
+
+This can be useful for tracking versions, categories, or any other relevant information.
 
 ```python
-prompt = BookRecommendationPrompt(genre="science")
-print(prompt.dump())
-# > {
-#     ...
-#     'inputs': {'genre': 'fantasy', , 'book_type': 'textbook'}
-#   }
+from mirascope.core import BasePrompt, prompt_template, metadata
+
+
+@metadata({"tags": {"version:0001", "category:books"}})
+@prompt_template("Recommend a {genre} book for a {age_group} reader.")
+class BookRecommendationPrompt(BasePrompt):
+    genre: str
+    age_group: str
+
+
+prompt = BookRecommendationPrompt(genre="fantasy", age_group="adult")
+print(prompt.dump()['metadata'])
+# > {'tags': {'version:0001', 'category:books'}}
 ```
 
-You should note that this is just one example, and you can use all Pydantic functionalities to improve your prompts. Check out their docs here[link].
+!!! note "Adding Additional Fields"
+
+    Although `Metadata` is a `TypedDict` with only the `tags` key, there is nothing stopping you from adding additional keys. The only issue is that this will throw a type error, which you can ignore. If there are particular keys you find yourself using frequently, let us know so we can add it!
```

Comment:
maybe we should have a `custom` or `user_defined` key that can catch-all user defined metadata so there isnt a type error?

Reply:
I thought about something like that, but at that point we should just make metadata `dict[str, Any]`

I opted against that because I like have the autocomplete on the `tags` key and figured a type ignore for additional keys until we add them is totally fine (and I couldn't really think of other non-tag metadata we should add at this moment so will leave that to users).

I can also make this a little more clear around adding something like `# pyright: ignore [reportArgumentType]` instead of just a catch all `# type: ignore`?

Reply:
yeah that seems like a good update

Reply:
Done.